From 27367e039656f4b1e6c8cecaf19c9eea458b0842 Mon Sep 17 00:00:00 2001
From: Prakash.Kumar <prakumar@habana.ai>
Date: Thu, 20 Jan 2022 05:44:14 +0200
Subject: [PATCH] [SW-65531] efuse: efuse data read during init

This patch adds support for reading the efuse data
in the first efuse initialization. This init reads
and stores complete efuse data in the psoc efuse
data register which works as shadow register. This
is done to avoid multiple physical access to efuse.
Also efuse init is avoided in rest of the boot flow.

Change-Id: I2751493297c0d57fffd482bda172c3c58ff8778e
Signed-off-by: Prakash.Kumar <prakumar@habana.ai>
---

diff --git a/lib/common/efuse.c b/lib/common/efuse.c
index dab173a..a6066b9 100644
--- a/lib/common/efuse.c
+++ b/lib/common/efuse.c
@@ -30,7 +30,8 @@
 
 int preboot_fuse_init(void)
 {
-	int i;
+	int i, offset;
+	u32 read_reg;
 
 	if (efuse_init == 1)
 		return -1;
@@ -47,6 +48,16 @@
 	if (i == FUSE_RETRIES)
 		return -2;
 
+	for (i = 0; i < FUSE_WORDS_PER_BANK; i++) {
+		/* read value from bank0 */
+		offset = (i * sizeof(u32));
+		read_reg = regmap_readl(efuse.data_0_reg + offset);
+
+		/* read value from bank1 and OR with value from bank0 */
+		offset = ((FUSE_WORDS_PER_BANK + i) * sizeof(u32));
+		read_reg |= regmap_readl(efuse.data_0_reg + offset);
+	}
+
 	efuse_init = 1;
 	return 0;
 }
